services/alchemy_vision/v1.js

```diff
@@ -14,64 +14,171 @@
         return;
       }
 
-      apikey = apikey || this.credentials.apikey;
+      // If it is present the newly provided user entered key takes precedence over the existing one. 
+      apikey = s_apikey || this.credentials.apikey;
+      this.status({}); 
 
       if (!apikey) {
+        this.status({fill:'red', shape:'ring', text:'missing credentials'});          
         var message ='Missing Alchemy API service credentials'; 
         node.error(message, msg);
         return;
       }
 
-      var alchemy = new AlchemyAPI(apikey);
-
+      // Check which single feature has been requested.
       var feature = config["image-feature"];
 
-      alchemy[feature](msg.payload, msg.alchemy_options || {}, function (err, response) {
-        if (err || response.status === "ERROR") { 
-          var message = 'Alchemy API request error: ' + (err ? err : response.statusInfo); 
-          node.error(message, msg);
-          return;
+      // Splice in the additional options from msg.alchemy_options
+      // eg. The user may have entered msg.alchemy_options = {knowledgeGraph: 1};     
+      var params = {};
+
+      for (var key in msg.alchemy_options) { params[key] = msg.alchemy_options[key]; }
+
+      // This is the callback after the call to the alchemy service.    
+      // Set up as a var within this scope, so it has access to node, msg etc. 
+      // in preparation for the Alchemy service action       
+      var actionComplete = function(err, keywords) {
+        if (err || keywords.status === 'ERROR') {
+          node.status({fill:'red', shape:'ring', text:'call to alchmeyapi vision service failed'}); 
+          console.log('Error:', msg, err);
+          node.error(err, msg);
         }
+        else {
+          msg.result = keywords[FEATURE_RESPONSES[feature]] || [];
+          msg.fullresult = {};
+          msg.fullresult['all'] = keywords;
+          node.send(msg); 
+        }        
+      }        
+      
+      // If the input is an image, need to stream the input in, giving time for the
+      // data to arrive, before invoking the service. 
+      if (imageCheck(msg.payload)) {
+        temp.open({suffix: '.' + fileType(msg.payload).ext}, function (err, info) {
+          if (err) {
+            this.status({fill:'red', shape:'ring', text:'unable to open image stream'});          
+            var message ='Node has been unable to open the image stream'; 
+            node.error(message, msg);
+            return;        
+          }  
+
+          stream_buffer(info.path, msg.payload, function () {
+            params['image'] = fs.createReadStream(info.path);
+            performAction(params, feature, actionComplete, temp.cleanup);        
+          });
+
+        });
+      } else if (urlCheck(msg.payload)) {
+        params['url'] = msg.payload;
+        performAction(params, feature, actionComplete);
+      } else {
+        this.status({fill:'red', shape:'ring', text:'payload is invalid'});          
```

Comment:
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Variable 'message' is already declared in line 124.](https://www.codacy.com/app/rezgui-y/node-red-node-watson/file/2569415486/issues/source?bid=3225115&fileBranchId=3297455#l177)